### ISSUE TITLE
Unnamed dock widget shows in plugin list as empty entry, but named is not present, 

### ISSUE THREAD

**Issue by Czaki** [napari/napari#3316]
## 🐛 Bug

![Zrzut ekranu z 2021-09-01 09-24-55](https://user-images.githubusercontent.com/3826210/131732523-9e8bd53a-a18f-41bb-95bd-05641030e2ef.png)

## To Reproduce

```python
from qtpy.QtWidgets import QLabel

import napari


viewer = napari.Viewer()

label1 = QLabel("text1")
label2 = QLabel("text2")

viewer.window.add_dock_widget(label1)
viewer.window.add_dock_widget(label2, name="label2")

napari.run()
```

The plugins menu will show an empty entry which will control `label1` visibility. 

## Expected behavior

Not showing controls for docked with `add_dock_widget` widget or show it for named ones.

## Environment

511db6e4fb16a6188754ae4ba94a17d2ecaab0da commit. 

 - Please copy and paste the information at napari info option in help menubar here:

 - Any other relevant information:

**Comment by tlambert03** [comment]
looks like this has always been the case (it just used to be a blank entry in the "Window" menu rather than the "plugins" menu).  I think we could/should provide nicer magic/default labeling there if `name` is not provided ... But the real bug now is that a dock widget added by an end user using `add_dock_widget` may actually end up in the `Plugins` menu (which should be strictly preserved for `add_plugin_dock_widget`.  Looks like that came in with #3190, which puts a lot of plugin-menu-specific logic into the generic `_add_viewer_dock_widget` method:

https://github.com/napari/napari/blob/8f1126c82bfbabd6cafd9739fa8c747ddff79583/napari/_qt/qt_main_window.py#L854-L888

**Comment by tlambert03** [comment]
cc @ppwadhwa ... could you have a look at this?  We don't want anything but dock widgets provided by plugins to get into the `Plugins` menu.  If an end-user uses `window.add_dock_widget(some_widget)` ... they shouldn't look in the plugins menu for it, it should always be the windows menu.  whereas `window.add_dock_widget(plugin_name)` is the programatic way to open up a plugin-provided dock widget (Which _should_ be in the plugins menu).
As for the empty name, that's a bug here:

https://github.com/napari/napari/blob/8f1126c82bfbabd6cafd9739fa8c747ddff79583/napari/_qt/qt_main_window.py#L753-L760

because `name = widget.objectName()` may still return an empty string (not always an `AttributeError`)... so we need to catch that case and give it the "Dock widget..." name

**Comment by sofroniewn** [comment]
> But the real bug now is that a dock widget added by an end user using add_dock_widget may actually end up in the Plugins menu (which should be strictly preserved for add_plugin_dock_widget. Looks like that came in with #3190, which puts a lot of plugin-menu-specific logic into the generic _add_viewer_dock_widget method

Ok yup, we should probably change. I think @ppwadhwa is out for the rest of this week. @goanpeca any chance you could pick this up if we want to fix before the release?

**Comment by goanpeca** [comment]
I cam take a look

**Comment by goanpeca** [comment]
@tlambert03 should the `add_dock_widget` just fail if the thing being added is not a dockwidget?

should there be a separate `add_widget` method? I find it confusing that adding a dock or a widget on a method called `add_dock_widget` would put things on different menus 🤔

**Comment by tlambert03** [comment]
> should the add_dock_widget just fail if the thing being added is not a dockwidget

this method takes a widget and creates a docked widget for the user.  Users are not expected to create QDockWidgets.  it currently does fail if it's not a widget since `QtViewerDockWidget(..., widget)` will raise an exception if it's not compatible.

> should there be a separate `add_widget` method? 

no?  this is the `add_widget`.  I think saying `window.add_widget` is fine if you want to change it (but that's an API break)(, `add_dock_widget` was used to suggest that the widget will be docked on the side of the window (not that it should already be a docked widget).

> I find it confusing that adding a dock or a widget on a method called add_dock_widget would put things on different menus 🤔

I do too!  that's the point here, and that's what got changed in #3190.  `add_dock_widget` is _not_ for plugins.  it's for users to add their own QWidget to the window.  It should _only_ put things in the Window menu.

**Comment by goanpeca** [comment]
@tlambert03 I am confused with

>it should always be the windows menu. whereas window.add_dock_widget(plugin_name) is the programatic way to open up a plugin-provided dock widget (Which should be in the plugins menu).

That methods always requires a widget, so how doe sthat work?

**Comment by tlambert03** [comment]
i'm not sure what you're asking... can you look at the source code (perhaps before #3190 if necessary) and see if you're still confused about the intention?

**Comment by tlambert03** [comment]
oh crap sorr ... that was a typo.  I used the same function name twice:

`window.add_dock_widget(some_widget: QWidget)` -> add a user-created widget to the window, put it in the `Window` menu

`window.add_plugin_dock_widget(plugin_name: str)` -> request that an already registered plugin widget be created and added to the window.

**Comment by goanpeca** [comment]
Ahhhh ok, :-p

**Comment by sofroniewn** [comment]
This got closed by #3326, I can reopen for / make a new issue for the following or just leave it as is if PR is coming soon :-)

> window.add_dock_widget(some_widget: QWidget) -> add a user-created widget to the window, put it in the Window menu
>
> window.add_plugin_dock_widget(plugin_name: str) -> request that an already registered plugin widget be created and added to the window.
